tv_apps/smart-system/js/sync_manager.js

```diff
@@ -0,0 +1,621 @@
+      // permanent 'syncing' state.
+      var killedApp = event.detail.origin;
+      navigator.mozApps.getSelf().onsuccess = event => {
+        var app = event.target.result;
+        app.getConnections().then(connections => {
+          connections.forEach(connection => {
+            if (connection.keyword != SYNC_REQUEST_IAC_KEYWORD ||
+                connection.subscriber != killedApp) {
+              return;
+            }
+            Service.request('SyncStateMachine:error', ERROR_SYNC_APP_KILLED);
+          });
+
+        });
+      };
+    },
+
+    /** Helpers **/
+
+    initStore: function() {
+      var promises = [];
+      [SYNC_STATE,
+       SYNC_STATE_ERROR,
+       SYNC_LAST_TIME].forEach(key => {
+        var promise = new Promise(resolve => {
+          asyncStorage.getItem(key, value => {
+            this.debug(key + ': ' + value);
+            this.store.set(key, value);
+            resolve();
+          });
+        });
+        promises.push(promise);
+      });
+      return Promise.all(promises);
+    },
+
+    updateState: function(state) {
+      this.state = state || Service.query('SyncStateMachine.state');
+    },
+
+    registerSyncRequest: function() {
+      return navigator.sync.registrations().then(registrations => {
+        var existingReg;
+        registrations.some(reg => {
+          if (reg.task === SYNC_TASK) {
+            return !!(existingReg = reg);
+          }
+        });
+
+        if (existingReg) {
+          this.debug('Sync request already exists');
+          return Promise.resolve();
+        }
+
+        this.debug('Registering sync request', location.href);
+
+        // XXX We need to default to the settings value (3600 and false)
+        // because of
+        // Bug 1206004 - BaseModule._settings is not be available during
+        //               BaseModule._start
+        var minInterval = this.DEBUG ? 120
+                          : this._settings['sync.scheduler.interval'] || 3600;
+        var wifiOnly = this._settings['sync.scheduler.wifionly'] || false;
+
+        return navigator.sync.register(SYNC_TASK, {
+          minInterval: minInterval, // minInterval is in seconds.
+          oneShot: true,
+          wifiOnly: wifiOnly,
+          data: null,
+          wakeUpPage: location.href
+        });
+      });
+    },
+
+    unregisterSyncRequest: function() {
+      navigator.sync.unregister(SYNC_TASK).catch(e => {
+        // A failure unregistering the sync task might be expected
+        // depending on the state we are (i.e. enabling). In any case,
+        // we show a warning, so a bug can be identified or discarded.
+        console.warn('Could not unregister sync task', e,
+                     Service.query('SyncStateMachine.state'));
+      });
+    },
+
+    getKeys: function() {
+      return new Promise((resolve, reject) => {
+        LazyLoader.load('js/fx_accounts_client.js', () => {
+          FxAccountsClient.getKeys(resolve, reject);
```

Comment:
You need https://bugzilla.mozilla.org/show_bug.cgi?id=1182071 to make this work.